cli_tools/gce_windows_upgrade/upgrader/workflows.go

```diff
@@ -0,0 +1,480 @@
+	w.Steps = map[string]*daisy.Step{
+		"restore-script": {
+			UpdateInstancesMetadata: &daisy.UpdateInstancesMetadata{
+				{
+					Instance: u.instanceURI,
+					Metadata: map[string]string{
+						metadataKeyWindowsStartupScriptURL:       u.getOriginalStartupScriptURL(),
+						metadataKeyWindowsStartupScriptURLBackup: "",
+					},
+				},
+			},
+		},
+		"detach-install-media-disk": {
+			DetachDisks: &daisy.DetachDisks{
+				{
+					Instance:   u.instanceURI,
+					DeviceName: daisyutils.GetDeviceURI(u.instanceProject, u.instanceZone, u.installMediaDiskName),
+				},
+			},
+		},
+		"delete-install-media-disk": {
+			DeleteResources: &daisy.DeleteResources{
+				Disks: []string{
+					daisyutils.GetDiskURI(u.instanceProject, u.instanceZone, u.installMediaDiskName),
+				},
+			},
+		},
+	}
+	w.Dependencies = map[string][]string{
+		"delete-install-media-disk": {"detach-install-media-disk"},
+	}
+	return nil
+}
+
+func (u *upgrader) rollback() (*daisy.Workflow, error) {
+	return u.runWorkflowWithSteps("rollback", u.Timeout, populateRollbackSteps)
+}
+
+func populateRollbackSteps(u *upgrader, w *daisy.Workflow) error {
+	stepStopInstance, err := daisyutils.NewStep(w, "stop-instance")
+	if err != nil {
+		return err
+	}
+	stepStopInstance.StopInstances = &daisy.StopInstances{
+		Instances: []string{u.instanceURI},
+	}
+
+	stepDetachNewOSDisk, err := daisyutils.NewStep(w, "detach-new-os-disk", stepStopInstance)
+	if err != nil {
+		return err
+	}
+	stepDetachNewOSDisk.DetachDisks = &daisy.DetachDisks{
+		{
+			Instance:   u.instanceURI,
+			DeviceName: daisyutils.GetDeviceURI(u.instanceProject, u.instanceZone, u.osDiskDeviceName),
+		},
+	}
+
+	stepAttachOldOSDisk, err := daisyutils.NewStep(w, "attach-old-os-disk", stepDetachNewOSDisk)
+	if err != nil {
+		return err
+	}
+	stepAttachOldOSDisk.AttachDisks = &daisy.AttachDisks{
+		{
+			Instance: u.instanceURI,
+			AttachedDisk: compute.AttachedDisk{
+				Source:     u.osDiskURI,
+				DeviceName: u.osDiskDeviceName,
+				AutoDelete: u.osDiskAutoDelete,
+				Boot:       true,
+			},
+		},
+	}
+
+	stepDetachInstallMediaDisk, err := daisyutils.NewStep(w, "detach-install-media-disk", stepAttachOldOSDisk)
+	if err != nil {
+		return err
+	}
+	stepDetachInstallMediaDisk.DetachDisks = &daisy.DetachDisks{
+		{
+			Instance:   u.instanceURI,
+			DeviceName: daisyutils.GetDeviceURI(u.instanceProject, u.instanceZone, u.installMediaDiskName),
+		},
+	}
+
+	stepRestoreScript, err := daisyutils.NewStep(w, "restore-script", stepDetachInstallMediaDisk)
+	if err != nil {
+		return err
+	}
+	stepRestoreScript.UpdateInstancesMetadata = &daisy.UpdateInstancesMetadata{
+		{
+			Instance: u.instanceURI,
+			Metadata: map[string]string{
+				metadataKeyWindowsStartupScriptURL:       u.getOriginalStartupScriptURL(),
+				metadataKeyWindowsStartupScriptURLBackup: "",
+			},
+		},
+	}
+
+	stepStartInstance, err := daisyutils.NewStep(w, "start-instance", stepRestoreScript)
```

Comment:
What if the instance was not running before the upgrade? We should leave it in the same state as before the upgrade while this appears to start the instance idiscriminately.

Reply:
customer needs to verify whether the instance's functionality (applications installed) works as expected - that is also explained in the guide. So I leave it "started". Would you think about it again?

Reply:
Customer's could start the instance if they want to do the verification. It's probably minor, but I feel we should keep the state of the instance unchanged.

Reply:
created b/156668741 to add an option to control the final state. By default, we set it to always stopped after discussion in order to save money for customer.